tinygrad/codegen/triton.py

```diff
@@ -0,0 +1,96 @@
+from typing import Callable, Dict, Final
+from triton.compiler import compile as triton_compile # type: ignore
+import hashlib
+import math
+
+from tinygrad.ops import BinaryOps, ASTRunner, FusedOps, Op, UnaryOps
+from tinygrad.codegen.linearizer import Linearizer, LocalBuffer, UOps
+from tinygrad.helpers import DEBUG, ImageDType, dtypes
+from tinygrad.shape.symbolic import NumNode
+
+class TritonCodegen(Linearizer):
+  def codegen(self):
+    self.process()
+    self.limit_global_dims(3)
+    self.linearize()
+
+    kernel = []
+    global_size = []
+    local_size = []
+    depth = 0
+    def kk(s): kernel.append("  "*depth+s)
+
+    kk("@triton.jit")
+    kk("def fxn("+','.join(f"data{i}" for i in range(len(self.bufs)))+"):")
+    depth += 1
+
+    gid = [f"tl.program_id({i})" for i in range(3)]
+    code_for_op: Final[Dict[Op, Callable]] = {
+      UnaryOps.EXP2: lambda x: f"tl.math.exp2({x})",
+      UnaryOps.LOG2: lambda x: f"tl.math.log2({x})", # TODO: is fast_log2f ok?
+      UnaryOps.SIN: lambda x: f"tl.sin({x})",
+      BinaryOps.ADD: lambda x,y: f"({x}+{y})", BinaryOps.SUB: lambda x,y: f"({x}-{y})",
+      BinaryOps.MUL: lambda x,y: f"({x}*{y})", BinaryOps.DIV: lambda x,y: f"({x}/{y})", # fdiv?
+      BinaryOps.POW: lambda x,y: f"tl.math.pow({x}, {y})", BinaryOps.MAX: lambda x,y: f"tl.maximum({x},{y})", # axis?
+      BinaryOps.CMPEQ: lambda x,y: f"({x}=={y})",
+      FusedOps.MULACC: lambda a,b,c: f"tl.math.fma({a}, {b}, {c})"
+    }
+    bufnames = ["temp" if isinstance(b, LocalBuffer) else f"data{i}" for i,b in enumerate(self.bufs)]
+
+    for uop,newvar,vin,args in self.uops:
+      if uop == UOps.LOOP:
+        for i,var in enumerate(args[0]):
+          if isinstance(var, NumNode): continue # python doesnt have block scope
+          else:
+            if args[1] == "global":
+              global_size.append(var.max+1)
+              kk(f"{var.expr} = {gid[i]} # {var.max+1}")
+            elif args[1] == "local":
+              local_size.append(var.max+1)
+              kk(f"{var.expr} = tl.arange({var.min}, {var.max+1})")
```

Comment:
This doesn't work, it only doesn't complain because there's no optimizer. Can you put back `hand_coded_optimizations` to match cstyle?

Reply:
Working on this, implementing local loops has caused some issues with block pointers in triton, and implementing `ReduceOps.SUM` has proved challenging, specifically finding the dimension to reduce along from just the output of the linearizer. Working on changing the linearizer to specify the axis in question.